### PR TITLE
feat: metrics functionality

### PR DESCRIPTION
**Pull request by asmie**
Metrics functionality which allow to create histograms, counters and other statistics used when measuring latencies, completion or other parameters.

### CODE REVIEW

File: ziggurat-core-metrics/src/recorder.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        } else {
+            None
+        }
+    }
+}
+
+pub struct TestMetrics(Snapshotter);
+
+impl TestMetrics {
+    fn new() -> TestMetrics {
```
Comment (ghost) [inline ziggurat-core-metrics/src/recorder.rs]: new() -> Self

File: ziggurat-core-metrics/src/recorder.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    }
+}
+
+impl Default for TestMetrics {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TestMetrics {
```
Comment (ghost) [inline ziggurat-core-metrics/src/recorder.rs]: Why multiple `impl TestMetrics` blocks?

File: ziggurat-core-metrics/src/recorder.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        histogram.record(1.0);
+
+        let histogram2 = register_histogram!(METRIC_NAME_ALT);
+        histogram2.record(1.0);
+
+        let snapshot = metrics.take_snapshot();
+        let constructed_histogram = snapshot.construct_histogram(METRIC_NAME).unwrap();
+        let constructed_histogram2 = snapshot.construct_histogram(METRIC_NAME_ALT).unwrap();
+
+        assert!(constructed_histogram.entries() == 1);
```
Comment (ghost) [inline ziggurat-core-metrics/src/recorder.rs]: assert_eq

**Review by ghost — APPROVED** [review]

File: ziggurat-core-metrics/src/recorder.rs (hunk 2 above)
Comment (asmie) [reply ziggurat-core-metrics/src/recorder.rs]: Merged.

**Pull request merged**